d2/src/main/java/com/linkedin/d2/xds/balancer/XdsDirectory.java

```diff
@@ -0,0 +1,150 @@
+package com.linkedin.d2.xds.balancer;
+
+import com.linkedin.common.callback.Callback;
+import com.linkedin.d2.balancer.Directory;
+import com.linkedin.d2.xds.XdsClient;
+import indis.XdsD2;
+import io.grpc.Status;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.linkedin.d2.xds.XdsClient.*;
+
+
+public class XdsDirectory implements Directory
+{
+  private final XdsClient _xdsClient;
+  private final ConcurrentMap<String, String> _serviceNames = new ConcurrentHashMap<>();
+  private final ConcurrentMap<String, String> _clusterNames = new ConcurrentHashMap<>();
+  private final AtomicReference<WildcardD2ClusterOrServiceNameResourceWatcher> _watcher = new AtomicReference<>();
+  /**
+   * A flag that shows whether the service/cluster names data is being updated. Requests to the data should wait until
+   * the update is done.
+   */
+  private final AtomicBoolean _isUpdating = new AtomicBoolean(true);
+  /**
+   * This lock will be released when the service and cluster names data have been updated and is ready to serve.
+   * If the data is being updated, requests to access the data will wait indefinitely. Callers could set a shorter
+   * timeout on the callback passed in to getServiceNames or getClusterNames, as needed.
+   */
+  private final Object _dataReadyLock = new Object();
```

Comment:
Using the `synchronized` block is making this more complicated than it needs to be I think. You could use a `ReadWriteLock` and do this in `onChanged` and `onRemoval`:
```java
if (_isUpdating.compareAndSet(false, true)) {
  _dataReadyLock.writeLock().lock();
}
```
And then the reverse in `onAllResourcesProcessed`:
```java
_isUpdating.set(false);
_dataReadyLock.writeLock().unlock();
```

Then in `waitAndRespond`:
```java
    List<String> names;
    try {
      _dataReadyLock.readLock().lock()
      names = new ArrayList<>(isForService ? _serviceNames.values() : _clusterNames.values());
    } finally {
      _dataReadyLock.readLock().unlock();
    }
    callback.onSuccess(names);
```

Reply:
smart. Updated.